skypilot-org/skypilot
# [Azure] Allow resource group specifiation for Azure instance provisioning

This resolves #3659 





Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
  - [x] Restart a stopped instance created with resource group specified.
  - [x] Relaunch on existing running cluster that was created with resource group specified.
  - [x] launch with `storage_account` along with `resource_group_vm` both specified at `~/.sky/config.yaml`
  - [x] Test Instance Deletion After Resource Group Switch
  1. create an instance, A, specifying one RG
  2. change the specified RG to another RG
  3. create another instance, B, with switched RG
  4. keep the RG used for B at `config.yaml` and remove A
  5. Check if A is removed successfully.
  - [x] Check if replicas with `sky serve up` is created at the specified resource group.
- [x] All smoke tests: `pytest tests/test_smoke.py` 
  - [x] `pytest tests/test_smoke.py --azure` except:
  1. `tests/test_smoke.py::test_cancel_pytorch`
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [x] Manual backward compatibility tests:
  - [x] Ensure proper handling of Azure instances across different branches with and without resource group specification
  1. Create an Azure instance from master branch using default resource group name
  2. switch to resource group specification branch
  3. create an instance with `resource_group_vm` specified
  4. delete the instance created from master branch
      - check if it removes the entire RG
  5. delete the instance created from RG specification.
      - check if it removes the VM, attached resources, deployments, and role assignment without removing resource group.
